### ISSUE TITLE
[RFC] HiCache stage 3: end-to-end integration and validation

### ISSUE THREAD

**Issue by cjx0709** [sgl-project/sglang-jax#1313]
## Summary

Wire the modules from Stage 0-2 (`HostKVPool`, `UnifiedRadixCache` + `FullComponent`, `HiCacheController`) end-to-end. This stage adds the complete **control plane** — the decision layer that determines when to transfer, which nodes to transfer, and how to handle resource exhaustion — on top of the Stage 2 controller (which only handles the "how to transfer" execution).

Specifically:
1. Replace `FullComponent` hook stubs with real implementations
2. Add control plane methods to `UnifiedRadixCache` (write_backup, load_back, eviction paths, lock management, async event management)
3. Integrate with the scheduler at three call sites
4. Validate correctness (KL equivalence) and performance (forward latency regression)

## Position in the roadmap

1. Host pool foundation
2. UnifiedRadixCache + FullComponent port
3. HiCacheController + async pipeline
4. **End-to-end integration and validation** *(this issue)*
5. L3 persistent storage
6. Multi-component + tuning

This issue depends on all three prior stages. Stage 4 (L3 storage ABC) can start in parallel.

## Proposed structure changes

- `sgl_jax/srt/mem_cache/unified_radix_cache.py` — **modify**: add `write_backup`, `load_back`, `writing_check`, `loading_check`, `check_hicache_events`, `flush_write_through_acks`, `init_load_back`, eviction paths (`_evict_device_leaf`, `_evict_to_host`, `_evict_host_leaf`, `evict_host`), leaf set management, lock mechanism
- `sgl_jax/srt/mem_cache/unified_cache_components/full_component.py` — **modify**: stub -> real hook implementation (`build_hicache_transfers`, `commit_hicache_transfer`, `evict_component`, `drive_eviction`, `drive_host_eviction`, `acquire_component_lock`)
- `sgl_jax/srt/mem_cache/kv_cache_builder.py` — **modify**: call `init_hicache()` in builder
- `sgl_jax/srt/managers/scheduler.py` — **modify**: add `check_hicache_events` + `init_load_back` in `get_new_batch_prefill`; add `flush_write_through_acks` in `update_running_batch`
- `sgl_jax/test/mem_cache/test_hicache_e2e.py` — **new**: end-to-end tests

### 1. Control plane overview

The control plane is distributed across three modules:

| Module | Responsibility |
|---|---|
| `UnifiedRadixCache` | Node-level D2H/H2D decisions, eviction paths, async event management |
| `FullComponent` | Resource release, heap-sorted eviction, lock management, transfer construction |
| `Scheduler` | Step-level scheduling entry points |

Stage 2 already provides the controller layer (`write`, `load`, `check_write_status`, `drain_pending`, `evict_callback`). Stage 3 adds the following control plane on top.

### 2. write_backup: D2H write decision

`write_backup` is the D2H **decision layer** (controller.write is the execution layer):

1. **Parent invariant** (Full attention only): under write_through, parent must be backed up before child, enforced recursively. This ensures prefetch can find complete host data along the parent chain. This invariant relies on Full attention's prefix semantics — SWA's window-external parent KV has no meaning, and linear attention has no traditional KV cache. Backup strategy must be re-evaluated when extending to these architectures.
2. **Host space pre-eviction**: when space is insufficient, call `evict_host()` to free host leaf nodes; abandon if still insufficient.
3. **Node locking**: `inc_lock_ref` after submission to prevent eviction during in-flight D2H. Lock released by `writing_check` upon completion.
4. **Trigger condition**: called by `_inc_hit_count` when `hit_count >= write_through_threshold` (Stage 2 §5.1).

### 3. load_back: H2D load decision

`load_back` is the H2D decision layer:

1. **Transfer construction**: walk up from `best_match_node` collecting host_indices of all evicted+backuped nodes.
2. **Device space pre-eviction**: evict device leaf nodes when device pool is insufficient; abandon if still insufficient. HiCache does not reserve device pool space — reservation relies on the scheduler's existing `new_token_ratio` mechanism (decode headroom).
3. **mem_quota check**: scheduler-provided memory quota limits single load size to prevent device pool exhaustion.
4. **Path locking**: `inc_lock_ref` locks the device path from root to best_match_node; `loading_check` releases after completion.
5. **Commit**: write device_indices back to node after H2D, clear host_value, mark as non-evicted.

### 4. Device eviction paths

`_evict_device_leaf` selects a path based on backup state:

| Condition | Path | Description |
|---|---|---|
| `backuped == True` | `_evict_to_host` | Release device resources, demote node to host-only (stays in tree) |
| `not backuped` + `write_through` | cascade delete | No backup and no remedy — completely remove from tree |

`_evict_to_host`: call `evict_component(target=DEVICE)` to free device slot -> mark `evicted = True` -> update leaf set.

`drive_eviction` (FullComponent): heap-sort from `evictable_device_leaves` by priority, evicting one by one until enough tokens are freed.

### 5. Host eviction path

Triggered when `host_pool.alloc()` returns None:

1. `drive_host_eviction` (FullComponent): heap-sort traverse `evictable_host_leaves`
2. `_evict_host_leaf`: atomically release host resources, remove node from tree, cascade-delete tombstone parents

**Leaf set management**: maintain `evictable_device_leaves` and `evictable_host_leaves` sets, updated after every eviction/load/backup.

| Type | Criteria |
|---|---|
| Device leaf | non-root, non-evicted, `lock_ref == 0`, no children holding device value |
| Host leaf | non-root, evicted, backuped, `host_lock_ref == 0`, no children |

### 6. Lock mechanism

Two-layer locking prevents node eviction during async transfers:

| Lock | Protects | Lock time | Release time |
|---|---|---|---|
| `lock_ref` | device KV data | write_backup submission, load_back path locking | `writing_check` / `loading_check` completion |
| `host_lock_ref` | host KV data | load_back host node locking | `loading_check` completion |

`acquire_component_lock`: walk from node up to root, skip evicted segments, `lock_ref += 1` for device-on nodes and remove from leaf set. First version H2D is synchronous, so `host_lock_ref` is not needed yet.

### 7. Async event management

On top of Stage 2's `check_write_status` (future poll), add tree state updates:

- **`writing_check`**: mark `node.backuped = True` on D2H completion, release `lock_ref`, update leaf set
- **`loading_check`**: release `lock_ref` on H2D completion

Poll uses `Future.done()` non-blocking check.

### 8. FullComponent hook implementation

Replace Stage 1 stub hooks with real implementations:

- **`build_hicache_transfers`**: BACKUP_HOST phase returns device_indices; LOAD_BACK phase walks parent chain collecting evicted nodes' host_indices
- **`commit_hicache_transfer`**: BACKUP_HOST writes host_value; LOAD_BACK restores device value and clears host_value
- **`evict_component`**: release device slot and/or host slot by target

### 9. Scheduler integration

```text
Scheduler.event_loop_normal
  +- get_next_batch_to_run
       +- get_new_batch_prefill
       |    +- [entry] check_hicache_events()  -> non-blocking poll D2H/H2D completion
       |    +- init_next_round_input -> match_prefix
       |    |    +- [if L2 hit] init_load_back -> load_back -> controller.load()
       |    +- [tail] ready_to_load_host_cache()  -> first version H2D is sync, no-op here
       +- update_running_batch
       |    +- flush_write_through_acks()  -> early release D2H locks, increase evictable space
       +- run_batch -> forward (D2H runs async in background)
```

`_inc_hit_count` -> `write_backup` -> `controller.write()` is triggered automatically by insert, no additional scheduler call needed.

### 10. Initialization assembly

`init_hicache()` is called in `kv_cache_builder.py` after creating `UnifiedRadixCache` (when `hicache_storage != "disable"`). It creates `HostKVPool` and `HiCacheController`, and wires them to the cache and FullComponent.

### 11. Stage 3 scope summary

| Component | Stage 2 provides | Stage 3 adds |
|---|---|---|
| D2H execution | `controller.write()` | — |
| D2H decision | — | `write_backup` (parent invariant, host pre-eviction, locking) |
| H2D execution | `controller.load()` | — |
| H2D decision | — | `load_back` (device pre-eviction, mem_quota, path locking, commit) |
| D2H poll | `check_write_status()` | `writing_check` (+ tree state update + lock release) |
| H2D poll | — | `loading_check` (lock release) |
| Device eviction | — | `_evict_device_leaf`, `_evict_to_host`, `drive_eviction` |
| Host eviction | `evict_callback` (slot free) | `_evict_host_leaf`, `drive_host_eviction`, leaf set management |
| Lock | — | `inc_lock_ref` / `dec_lock_ref`, `acquire_component_lock` |
| FullComponent hooks | stub | `build_hicache_transfers`, `commit_hicache_transfer`, `evict_component` |
| Scheduler entry | — | `check_hicache_events`, `flush_write_through_acks`, `init_load_back` |
| Init | — | `init_hicache` + builder integration |

## Testing

Stage 1 covers basic radix cache correctness. Stage 3 focuses on HiCache-specific scenarios; the full test matrix will be refined during implementation.

### KL equivalence test

| Config | Value |
|---|---|
| Model | Qwen3-8B |
| Hardware | v6e-1 |
| Control | Existing `RadixCache` (no HiCache) |
| Experiment | `UnifiedRadixCache` + `--hicache-storage none` |
| Metric | All output token logits KL divergence < 1e-6 |
| Test set | Multi-turn conversation (triggers prefix reuse + eviction + prefetch) |

### HiCache-specific test scenarios

| Scenario | Validation point |
|---|---|
| L1 eviction -> L2 hit -> prefetch | H2D correctly restores KV, skips prefix forward |
| L1 + L2 both evicted -> full miss | Falls back to complete forward, results correct |
| write_through trigger condition | Only nodes with `hit_count >= threshold` trigger D2H |

### Performance validation

| Metric | Threshold |
|---|---|
| Single step forward latency increase | < 2% |
| Measurement | 100 step average (excluding first XLA compilation) |
| Hardware/model | Qwen3-8B @ v6e-1 |

## Acceptance criteria

- KL equivalence: Qwen3-8B multi-turn conversation, logits KL < 1e-6
- Latency regression: forward latency increase < 2%
- HiCache-specific test scenarios pass
- No memory leak: host pool `available_size()` stable during sustained operation
- write_through correctness: only nodes with `hit_count >= threshold` trigger D2H
